crypto/dh/dh_ameth.c

```diff
@@ -566,34 +567,39 @@ static void *dh_pkey_export_to(const EVP_PKEY *pk, EVP_KEYMGMT *keymgmt)
         || !ossl_param_bld_push_BN(&tmpl, OSSL_PKEY_PARAM_FFC_G, g))
         return NULL;
 
-    if (q != NULL) {
-        if (!ossl_param_bld_push_BN(&tmpl, OSSL_PKEY_PARAM_FFC_Q, q))
-            return NULL;
-    }
-
-    /*
-     * This may be used to pass domain parameters only without any key data -
-     * so "pub_key" is optional. We can never have a "priv_key" without a
-     * corresponding "pub_key" though.
-     */
-    if (pub_key != NULL) {
-        if (!ossl_param_bld_push_BN(&tmpl, OSSL_PKEY_PARAM_DH_PUB_KEY, pub_key))
-            return NULL;
+    if (!domainparams) {
+        if (q != NULL) {
+            if (!ossl_param_bld_push_BN(&tmpl, OSSL_PKEY_PARAM_FFC_Q, q))
+                return NULL;
+        }
 
-        if (priv_key != NULL) {
-            if (!ossl_param_bld_push_BN(&tmpl, OSSL_PKEY_PARAM_DH_PRIV_KEY,
-                                        priv_key))
+        /*
+         * This may be used to pass domain parameters only without any key
+         * data - so "pub_key" is optional. We can never have a "priv_key"
+         * without a corresponding "pub_key" though.
+         */
+        if (pub_key != NULL) {
```

Comment:
So if `domainparams` is set to false, then surely `pub_key` should never be NULL? Its an error if it is.

Reply:
From the `EVP_get0_key` manual (Italic bold mine):

> To get the public and private key values use the **DH_get0_key()** function. A pointer to the public key will be stored in **\*pub_key**, and a pointer to the private key will be stored in **\*priv_key**. Either may be NULL if they have not been set yet, although ***if the private key has been set then the public key must be***. The values point to the internal representation of the public key and private key values. This memory should not be freed directly.  Any of the out parameters **pub_key** and **priv_key** can be NULL, in which case no value will be returned for that parameter.

Do you remember that we discussed this once already?

Reply:
I'm somehow missing your point?

Reply:
Hmmm, so am I.  As a matter of fact, I suggest you take it up with whoever wrote [this PR](https://github.com/openssl/openssl/pull/9753) :wink:, that's where this was introduced

Reply:
It makes sense in that PR because keys could be empty except for domain parameters. In this PR you are splitting them up. Therefore it no longer makes sense to send a *key* without at least the pub key set.

Reply:
I must be daft, I suddenly see what you see...

Reply:
That comment is *really* confusing, in hindsight